### PR TITLE
[dagster-k8s] fix: Prevent job hang when Pod exits before init

### PR DESCRIPTION
**Pull request by MattyKuzyk**
## Summary & Motivation
This is meant to address https://github.com/dagster-io/dagster/issues/22453

Previously, if a pod was Evicted by the Kubelet or force deleted before initialization was finished, the `wait_for_pod` function would simply wait forever, forcing us to kill the job.

To solve this, we added a wait timeout loop specifically for the initial pod creation. As soon as the pod is "seen" for the first time by the Kubernetes API, we break out of that loop and continue to the main logic loop of waiting for pod readiness.

If at any point in that second loop the pod no longer exists, we know that it was unexpectedly deleted, and thus we need to raise an error so that the job doesn't hang indefinitely.

## How I Tested These Changes
- Added unit tests 
- Ran successful jobs with these changes
- Killed pod on startup and confirmed that an error was raised, causing Dagster to retry the pod

```
$ python -m pytest python_modules/libraries/dagster-k8s/dagster_k8s_tests/unit_tests/test_client.py
WARNING:root:Unable to detect CI environment.  No test analytics will be sent.
=========================================================================================================== test session starts ============================================================================================================
platform darwin -- Python 3.10.3, pytest-7.4.3, pluggy-1.3.0
rootdir: /Users/matthew.kuzyk/Code/Repo/mattk-dagster
configfile: pyproject.toml
plugins: syrupy-3.0.6, dependency-0.5.1, Faker-18.4.0, typeguard-4.1.5, mock-3.3.1, requests-mock-1.11.0, cov-2.10.1, xdist-3.3.1, buildkite-test-collector-0.1.7, anyio-4.2.0, rerunfailures-10.0
collected 35 items

python_modules/libraries/dagster-k8s/dagster_k8s_tests/unit_tests/test_client.py ...................................                                                                                                                 [100%]

--------------------------------------------------------------------------------------------------------- snapshot report summary ----------------------------------------------------------------------------------------------------------

====================================================================================================== 35 passed in 151.12s (0:02:31) ======================================================================================================
```
## Changelog

Insert changelog entry or "NOCHANGELOG" here.

- [ ] `NEW` _(added new feature or capability)_
- [x] `BUGFIX` _(Fixed job hang when Pod exits before init)_
- [ ] `DOCS` _(added or updated documentation)_

### CODE REVIEW

**Comment by MattyKuzyk** [comment]
Is there anything wrong with migrating to fstrings? I noticed their use is rampant elsewhere in the codebase so I assume it's fine.

**Comment by yuhan** [comment]
@smackesey @danielgafni can you take a look?

**Comment by smackesey** [comment]
@yuhan My k8s is weak, I think I'm not the right person to review this

**Comment by yuhan** [comment]
adding @gibsondan (who recently made changes to k8s pipes) @alangenfeld (who built k8s pipes orginally)

**Review by danielgafni — CHANGES_REQUESTED** [review]
Amazing job! 

The logic seems correct to me.

Asked for a few small changes

File: python_modules/libraries/dagster-k8s/dagster_k8s/client.py [hunk 1]
```diff
@@ -525,21 +525,25 @@ def get_pod_names_in_job(self, job_name, namespace):
...
+    ) -> None:
         """Wait for a pod to launch and be running, or wait for termination (useful for job pods).
 
         Args:
+        ----
             pod_name (str): Name of the pod to wait for.
             namespace (str): Namespace in which the pod is located.
             wait_for_state (WaitForPodState, optional): Whether to wait for pod readiness or
                 termination. Defaults to waiting for readiness.
+            wait_initial_timeout (numeric, optional): Timeout after which to give up and raise exception
```
[the remark this answers is not in the document]
Comment (MattyKuzyk) [reply python_modules/libraries/dagster-k8s/dagster_k8s/client.py]: Done

File: python_modules/libraries/dagster-k8s/dagster_k8s/client.py [hunk 2]
```diff
@@ -588,7 +595,20 @@ def wait_for_pod(
...
                 continue
+            break
 
+        while True:
+            pods = self.core_api.list_namespaced_pod(
+                namespace=namespace, field_selector=f"metadata.name={pod_name}"
+            ).items
+            pod = pods[0] if pods else None
+            if pod is None:
+                raise DagsterK8sError("Pod was unexpectedly killed")
```
[the remark this answers is not in the document]
Comment (MattyKuzyk) [reply python_modules/libraries/dagster-k8s/dagster_k8s/client.py]: Done

**Comment by danielgafni** [comment]
@MattyKuzyk could you please run `make ruff`? Otherwise looking good!

**Comment by MattyKuzyk** [comment]
Ran `make ruff`, should be good to go.

**Review by danielgafni — APPROVED** [review]
LGTM unless @alangenfeld has some comments

**Comment by alangenfeld** [comment]
I trust y'all - not seeing any issues doing a quick skim

**Pull request merged**